### DISCUSSION TITLE
Making large quantities of new IfcPropertySet(s) is too heavy

### DISCUSSION THREAD

**Discussion by RyanDaley** [GeometryGym/GeometryGymIFC#85] [Q&A] [answered]
I am trying to make a large quantitiy of unique IfcPropertySet(s), and GeometryGym just seems to slow to handle the operation.  For instance, in a given IFC file, there is very likely tens of thousands of IfcEntity(s).  If I want to make a new IfcPropertySet that is applied to all objects, the operation is too heavy for GeometryGym.  Perhaps there is a better way to do this given GeometryGym's logics, but this is the code I have written:

```
List<GeometryGym.Ifc.IfcProperty> singleValueList = new List<GeometryGym.Ifc.IfcProperty>();
foreach (PropertyItem propertyItem in pset)
{
    singleValueList.Add(new GeometryGym.Ifc.IfcPropertySingleValue(db, propertyItem.IfcPropertyName, value)); 
}
var Pset = new GeometryGym.Ifc.IfcPropertySet(pset.Key, singleValueList);
for (int i = 0; i < objectsByGuid.Count(); i++)
{
    Pset.RelateObjectDefinition(objectsByGuid.ElementAt(i));
}
```

The slow part of the operation is adding a reference to the IfcPropertySet for each IfcEntity by doing the .RelatObjectDefinition() method for each individual IfcEntity.  This is the only way I could think to do this because there is no IfcPropertySet constructor that has an IEnumerable<IfcObjectDefinition> as an argument.  Perhaps there should be a constructor that allows multiple related objects to be set.  
@jmirtsch , do you know of a solution to this problem?

**Comment by RyanDaley** [comment, marked as answer]
Further, what I ultimately want to do is replace an IfcPropertySet for each IfcEntity with a new version of the IfcPropertySet with additional IfcProperty(s) and different values.  So, first I need to remove the reference to the origianl IfcPropertySet before creating the new IfcPropertySet with the new name.  This operation is also incredibly heavy because I am iterating through each IfcEntity in the Project.

```
for (int i = 0; i < objectsByGuid.Count(); i++)
{
    var RelDefines = db.OfType<GeometryGym.Ifc.IfcRelDefinesByProperties>().Where(x => x.RelatedObjects.Contains(objectsByGuid.ElementAt(i))).FirstOrDefault(y => y.RelatingPropertyDefinition.Any(z => z.Name == pset.Key));
    if (RelDefines != null)
    {
        RelDefines.RelatedObjects.Remove(objectsByGuid.ElementAt(i));
    }
}
```

**Reply by jmirtsch** [reply]
Hi Ryan,

Thanks for posting.  From the first post, you can simply create a new relationship relating a property set to a collection of entities.
Instead of your for loop enclosing this action ( Pset.RelateObjectDefinition(objectsByGuid.ElementAt(i)); )
you could replace the for loop with this
	new IfcRelDefinesByProperties(objectsByGuid, Pset);

Note that IfcTypeObjects have a direct attribute for property sets and don't use this relationship.  The relationship is for instances.
If you are modifying an existing file, then you will likely need to remove the existing property set.
I'd suggest code that will perform better in removing an existing property set association could be this:

	foreach (var element in objectsByGuid)
			{
				IfcRelDefinesByProperties definesByProperties = element.IsDefinedBy.Where(x => x.RelatingPropertyDefinition.Any(x => x.Name == pset.Key)).FirstOrDefault();
				if (definesByProperties != null)
				{
					if (definesByProperties.RelatedObjects.Count == 1)
						definesByProperties.Dispose(true);
					else
						definesByProperties.RelatedObjects.Remove(element);
				}
			}

Iterating the db is a very slow action.

Hope this helps,

Jon

**Reply by RyanDaley** [reply]
@jmirtsch ,

First, thank you for your reply. 

> 
> Iterating the db is a very slow action.
> 
Yea, so this makes sense that any call to the db is a slow action.  It is clear that modifications to the db are the bottleneck and should be minimized as much as possible in my code.     

Thanks for the help.
